Brokerages/Bitfinex/BitfinexBrokerage.Messaging.cs

```diff
@@ -0,0 +1,522 @@
+                var symbol = _symbolMapper.GetLeanSymbol(channel.Symbol);
+                var orderBook = _orderBooks[symbol];
+
+                var price = decimal.Parse(entries[0], NumberStyles.Float, CultureInfo.InvariantCulture);
+                var count = int.Parse(entries[1]);
+                var amount = decimal.Parse(entries[2], NumberStyles.Float, CultureInfo.InvariantCulture);
+
+                if (count == 0)
+                {
+                    orderBook.RemovePriceLevel(price);
+                }
+                else
+                {
+                    if (amount > 0)
+                    {
+                        orderBook.UpdateBidRow(price, amount);
+                    }
+                    else if (amount < 0)
+                    {
+                        orderBook.UpdateAskRow(price, amount);
+                    }
+                }
+            }
+            catch (Exception e)
+            {
+                Log.Error(e);
+                throw;
+            }
+        }
+
+        private void OnOrderClose(string[] entries)
+        {
+            string brokerId = entries[0];
+            if (entries[5].IndexOf("canceled", StringComparison.OrdinalIgnoreCase) >= 0)
+            {
+                var order = CachedOrderIDs
+                    .FirstOrDefault(o => o.Value.BrokerId.Contains(brokerId))
+                    .Value;
+                if (order == null)
+                {
+                    order = _algorithm.Transactions.GetOrderByBrokerageId(brokerId);
+                    if (order == null)
+                    {
+                        // not our order, nothing else to do here
+                        return;
+                    }
+                }
+                Order outOrder;
+                if (CachedOrderIDs.TryRemove(order.Id, out outOrder))
+                {
+                    OnOrderEvent(new OrderEvent(order, DateTime.UtcNow, 0, "Bitfinex Order Event") { Status = OrderStatus.Canceled });
+                }
+            }
+            else
+            {
+                lock (closeLocker)
+                {
+                    _pendingClose.Add(brokerId);
+                }
+            }
+        }
+
+        private void OnTradeExecuted(string[] entries)
+        {
+            try
+            {
+                var brokerId = entries[4];
+                var order = CachedOrderIDs
+                    .FirstOrDefault(o => o.Value.BrokerId.Contains(brokerId))
+                    .Value;
+                if (order == null)
+                {
+                    order = _algorithm.Transactions.GetOrderByBrokerageId(brokerId);
+                    if (order == null)
+                    {
+                        // not our order, nothing else to do here
+                        return;
+                    }
+                }
+                Trade trade = Trade.FromTU(entries);
+                EmitTradeTick(order, trade);
```

Comment:
I think we should move this line before the order checks, otherwise we'll only emit Trade ticks for our fills.

Reply:
Do mean that we still need Emit Trade Tick for order which is created outside LEAN?

Reply:
We should emit _Order Fill Events_ only for orders handled by LEAN and this is handled fine in the current code.

We should emit _Trade Ticks_ for **all** market trades, not only ours.